### ISSUE TITLE
The linemerge implementation is rather basic, more advanced version.

### ISSUE THREAD

**Issue by theomega** [abey79/vpype#173]
I'm using the `linemerge` operation in combination with `splitall` a lot to construct a new SVG with new paths hoping that there are as little as possible paths. I'm often not 100% satisfied with the results. Often there would be a solution to capture most/all points in a single path, but the implementation rather only finds small sub-paths.

After looking at the code, I think the current implementation is rather basic:
- It randomly picks a line which has not been processed yet
- Repeats the following:
  - Takes the end point of the current line and searching for another line with the closest starting point within the tolerance
  - If such a line is found, the two lines are joined
  - If no such line can be found, break the repeat

Especially if you have a network of paths (where from a single point is involved in multiple paths), this algorithm is often not finding the optimal solution.

Although, an optimal solution is up for definition. Would it be optimal to have the least amount of paths? Or optimize for having as long as possible paths? Or optimize for paths with as little pen-up-travel as possible?

Every implementation would most likely be a lot more complex (both with regards to run-time but also with regards to implementation), so I'm not sure if this is something which vpype wants do do better.

I don't have a good solution at hand, but I wanted to get the discussion started. Maybe someone has already found some good algorithms for this? Or maybe I'm wrong and doing something wrong.

Example where the linemerge does not achieve great results:
- Input file: https://gist.github.com/29d128cfa1981bc7f0fe7f19934a6a6e
- Vypye only reducing the file down to 2300 paths where there is most likely (but I might be wrong here) actually a version with a single path.
```
vpype read lion_0.1.svg stat splitall stat linemerge stat
========= Stats ========= 
Current page size: (1998.0, 1021.0)
Layer 1
  Length: 153819.32249758646
  Pen-up length: 17457.144205914872
  Total length: 171276.46670350133
  Mean pen-up length: 5.61142533137733
  Median pen-up length: 3.8479002962654754
  Path count: 3112
  Segment count: 21497
  Mean segment length: 7.155385518797342
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
Totals
  Layer count: 1
  Length: 153819.32249758646
  Pen-up length: 17457.144205914872
  Total length: 171276.46670350133
  Path count: 3112
  Segment count: 21497
  Mean segment length: 7.155385518797342
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
========================= 
========= Stats ========= 
Current page size: (1998.0, 1021.0)
Layer 1
  Length: 153819.32249758742
  Pen-up length: 17457.144205914872
  Total length: 171276.4667035023
  Mean pen-up length: 0.8121112860957793
  Median pen-up length: 0.0
  Path count: 21497
  Segment count: 21497
  Mean segment length: 7.155385518797386
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
Totals
  Layer count: 1
  Length: 153819.32249758742
  Pen-up length: 17457.144205914872
  Total length: 171276.4667035023
  Path count: 21497
  Segment count: 21497
  Mean segment length: 7.155385518797386
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
========================= 
========= Stats ========= 
Current page size: (1998.0, 1021.0)
Layer 1
  Length: 153824.25821345355
  Pen-up length: 59583.81956929929
  Total length: 213408.07778275284
  Mean pen-up length: 25.917276889647365
  Median pen-up length: 19.6467207808326
  Path count: 2300
  Segment count: 40694
  Mean segment length: 3.780023055326425
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
Totals
  Layer count: 1
  Length: 153824.25821345355
  Pen-up length: 59583.81956929929
  Total length: 213408.07778275284
  Path count: 2300
  Segment count: 40694
  Mean segment length: 3.780023055326425
  Bounds: (641.9081, 53.0234, 1181.3106, 730.4076)
========================= 
```

**Comment by theomega** [comment]
Here are some scientific reads:
Finding even the longest simple path between two points in a graph (so it contains a vertex at most once) is already NP Hard in general according to [wikipedia](https://en.wikipedia.org/wiki/Longest_path_problem). That sounds like finding the real optimal solution will be indeed, as I was mentioning computationally intense...

**Comment by theomega** [comment]
Yep, I get that and I completely agree with your approach.

With some pen plotters, the quality might improve if you have less pen-up and pen-downs and less travel. So it might not only be about printing time but also about quality...

**Comment by abey79** [comment]
That's a good point, yes. Let's see if someone would be willing to chime in and help on this front.

**Comment by tatarize** [comment]
The correct way to optimize the lines is implementation of two-opt, or if you wanted something a bit more k-opt Lin–Kernighan TSP solver.

The naive solution is actually pretty bad compared to the pareto-optimal solutions. The basic outline for 2-opt is that you take the current path you're using and you try swaps within it. You pick whatever two points you want, usually systematically. And check if you swapped that path between those points how would that change the end result? 

You would have four points for the swap location, you can then perform a cross at that location, this would actually reverse the entire path between these two points but for most metrics this changes nothing if we're highly concerned for travel distance. The only thing that changes is that our points A->B and a->b are now A->b and a->B and all the other points between these are now backwards but we can ignore that because the travel time is still the same backwards and forwards.

We then go down the line and do all the points, sometimes until we can show that no swap within the entire path can improve our results. Which just takes a couple passes. This might not be the optimal solution, but it's really quite close. And is pareto-optimal which is to say no single change can shorten our path, only lengthen our path.

We also, can do fancier things with this. While travel distance is a pretty good routine, there are other things we can do here. We can assign other metrics which weight certainly paths more than others. For example if we want to minimize direction changes. We can give weight to lines that are along the same general movement from each other. If we want to prioritize certain groups before other groups etc, you can put that into the weights and it'll optimize that, though in some cases the reverse of a path will weigh differently and then we need a bit more processing to finish that.

**Comment by tatarize** [comment]
Only a particular number of nodes will be within merging range. In theory that'll give you a set number of paths by definition. If you have 34 odd numbered nodes, you must by definition have 17 different paths. Since odd numbered nodes must start a path or end a path. You will therefore always have the same number of paths, though their lengths can absolutely vary.

**Comment by tatarize** [comment]
> Vypye only reducing the file down to 2300 paths where there is most likely (but I might be wrong here) actually a version with a single path.

No. 2300 likely is the minimum number of paths. Basically it means there are 4600 nodes of an odd order. There are absolutely more than 2 nodes with an odd number of attached segments therefore there can be no single walk which solves the entire graph. This is not a Eulerian graph. https://en.wikipedia.org/wiki/Seven_Bridges_of_K%C3%B6nigsberg

**Comment by tatarize** [comment]
That said there is a pretty easy algorithm to make a non-eulerian graph into a eulerian graph. Namely double it. Then all your nodes are even and by definition there is a single walk that reaches every node. If you wanted to go over the same graph twice, for any contiguous graph you could do it without ever lifting your pen.
